main.cpp

```diff
@@ -7,42 +7,371 @@
+		row = st_row;
+		col = st_col;
+		color = st_color;
+		row_turn = col_turn = 0;
+	}
+	virtual bool check_turn(int dst_row, int dst_col, char dst_color) {
+	/*	if (dst_row < col) col_turn = -col_turn;
+		if (dst_row < row) row_turn = -row_turn;
+		for (int i = col_turn, j = row_turn; col_turn + i < dst_col; i += col_turn, j += row_turn) {
+			if (field[row + j][col + i]->name != "0") return 0;
+		}
+		return 1;*/
+		return true;
+		// TODO: make check based on row_turn, col_turn
+	}
+	char get_color() const {
+		return color;
+	}
+	char get_row() const {
+		return row;
+	}
+	char get_col() const {
+		return col;
+	}
+	void change_row(int row_n) {
+		row = row_n;
+	}
+	void change_col(int col_n)  {
+		col = col_n;
+	}
+	virtual char to_char() const = 0;
 };
 
 class Pawn : public Figure {
 public:
-	bool check_turn() {
-		// ...
+	Pawn(int st_row, int st_col, char st_color) : Figure(st_row, st_col, st_color) {
+		// TODO: define row_turn, col_turn
+	}
+	bool check_turn(int dst_row, int dst_col, char dst_color) {
+		if (get_col() == dst_col && get_row() == dst_row) { cout << "You do not move" << endl; return 0; }
+		if (get_color() == dst_color) {cout << "There are your figure" << endl; return 0;}
+		if (get_color() == 'b') {
+			if (!(dst_col < 7 && get_col() >= 0 && dst_row < 8 && get_row() >= 0)) {
+				return false;
+			}
+			if (get_col() == dst_col) {
+				if (dst_row - get_row() == 1) {
+//					if (dst_row == 0) become_new_figure();
+					return true;
+				}
+				if (get_row() == 1 && dst_row == 3) {
+//					if (field[2][col]->color == "empty") {
+//						if (dst_row == 0) become_new_figure();
+						return true;
+	//				}
+				}
+			}
+			if (abs(dst_col - get_col()) == 1 && dst_row - get_row() == 1 && dst_color == 'w') {
+				return true;
+			}
+		}
+		if (get_color() == 'w') {
+			if (!(dst_col < 8 && get_col() > 0 && dst_row < 8 && get_row() > 0)) {
+				return false;
+			}
+			if (get_col() == dst_col) {
+				if (get_row() - dst_row == 1) {
+	//				if (dst_row == 7) become_new_figure();
+					return true;
+				}
+				if (get_row() == 6 && dst_row == 4) {
+	//				if (field[5][col]->get_color() == "empty") {
+	//					if (dst_row == 7) become_new_figure();
+						return true;
+//					}
+				}
+			}
+			if (abs(dst_col - get_col()) == 1 && get_row() - dst_row == 1 && dst_color == 'b') {
+				return true;
+			}
+		}
+		cout << "the move is not correct" << endl;
+		return 0;
+	}
+	char to_char() const {
+		return 'P';
+	}
+};
+class Rook : public Figure {
+public:
+	Rook(int st_row, int st_col, char st_color) : Figure(st_row, st_col, st_color) {}
+	bool check_turn(int dst_row, int dst_col, char dst_color) const {
+		if (get_col() == dst_col && get_row() == dst_row) { cout << "You do not move"; return 0; }
+		if (get_color() == dst_color) { cout << "There are your figure" << endl; return 0; }
+		if (!(dst_col < 8 && get_col() >= 0 && dst_row < 8 && get_row() >= 0)) {
+			cout << "the move is not correct" << endl;
+			return false;
```

Comment:
This whole block above is copy-pasted in all your classes. Thus, it makes sense to either move it to make_turn function of make a separate function just for this check and call it in the beginning of each of the class check_turn's. The reason why you might want to avoid code duplication is because if you need to make some change to this 5 lines, you will have to go to all places they are in the code and copy that change there. But if they are isolated in a single function, you can just modify it there and everything will be propagated.